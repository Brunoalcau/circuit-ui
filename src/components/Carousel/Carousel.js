/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { isEmpty, isFunction } from 'lodash/fp';
import {
  childrenPropType,
  childrenRenderPropType
} from '../../util/shared-prop-types';

import useComponentSize from './hooks/use-component-size';
import Container from './components/Container';
import Slides from './components/Slides';
import Slide from './components/Slide';
import SlideImage from './components/SlideImage';
import Controls from './components/Controls';
import Status from './components/Status';
import Progress from './components/Progress';
import Buttons from './components/Buttons';
import Step from '../Step';

import {
  ASPECT_RATIO,
  ANIMATION_DURATION,
  SLIDE_DURATION,
  SLIDE_DIRECTIONS
} from './constants';

const { ButtonList, NextButton, PrevButton, PlayButton } = Buttons;

const statusAlignment = ({ theme }) => css`
  flex: none;
  margin-right: ${theme.spacings.exa};

  ${theme.mq.untilKilo} {
    margin-right: ${theme.spacings.kilo};
  }
`;
const StyledStatus = styled(Status)(statusAlignment);

const progressAlignment = css`
  flex: 1 1 auto;
`;
const StyledProgress = styled(Progress)(progressAlignment);

const buttonsAlignment = ({ theme }) => css`
  margin-left: ${theme.spacings.exa};

  ${theme.mq.untilKilo} {
    margin-left: ${theme.spacings.kilo};
  }
`;
const StyledButtonList = styled(ButtonList)(buttonsAlignment);

const Carousel = ({
  slides,
  slideDuration,
  animationDuration,
  aspectRatio,
  children,
  cycle,
  swipe,
  autoPlay,
  pauseOnHover,
  hideControls,
  ariaLabelledbyPrefix,
  ...props
}) => {
  const slidesTotal = slides.length;
  const slidesRef = useRef(null);
  const slideSize = useComponentSize(slidesRef);
  const [slideDirection, setSlideDirection] = useState();

  if (isEmpty(slides)) {
    return null;
  }

  const goToNextSlide = () => setSlideDirection(SLIDE_DIRECTIONS.FORWARD);
  const goToPreviousSlide = () => setSlideDirection(SLIDE_DIRECTIONS.BACK);

  return (
    <Step
      cycle={cycle}
      swipe={swipe}
      autoPlay={autoPlay}
      pauseOnHover={pauseOnHover}
      totalSteps={slidesTotal}
      stepDuration={slideDuration}
      animationDuration={animationDuration}
      onNext={goToNextSlide}
      onPrevious={goToPreviousSlide}
    >
      {({
        step,
        previousStep,
        paused,
        play,
        pause,
        getStepProps,
        getNextControlProps,
        getPreviousControlProps,
        getPlayControlProps,
        getPauseControlProps
      }) => (
        <Container aria-label="gallery" {...props}>
          <Slides ref={slidesRef} {...getStepProps()}>
            {slides.map(({ image }, index) => (
              <Slide
                key={index}
                index={index}
                step={step}
                prevStep={previousStep}
                slideSize={slideSize}
                slideDirection={slideDirection}
                animationDuration={animationDuration}
              >
                <SlideImage
                  src={image.src}
                  alt={image.alt}
                  aspectRatio={aspectRatio}
                  aria-labelledby={
                    ariaLabelledbyPrefix
                      ? `${ariaLabelledbyPrefix}--${index}`
                      : undefined
                  }
                />
              </Slide>
            ))}
          </Slides>

          {!hideControls && (
            <Controls>
              <StyledStatus step={step} total={slidesTotal} />

              <StyledProgress
                key={step}
                paused={paused}
                animationDuration={slideDuration}
              />

              <StyledButtonList>
                <PlayButton
                  paused={paused}
                  {...(paused ? getPlayControlProps() : getPauseControlProps())}
                />
                <PrevButton {...getPreviousControlProps()} />
                <NextButton {...getNextControlProps()} />
              </StyledButtonList>
            </Controls>
          )}

          {isFunction(children)
            ? children({
                step,
                previousStep,
                paused,
                play,
                pause
              })
            : children}
        </Container>
      )}
    </Step>
  );
};

Carousel.propTypes = {
  slides: PropTypes.arrayOf(
    PropTypes.shape({
      image: PropTypes.object.isRequired
    })
  ),
  /**
   * Indicates duration of animation between slides (in milliseconds).
   */
  animationDuration: PropTypes.number,
  /**
   * Indicated time how long each slide will stay visible (in milliseconds).
   */
  slideDuration: PropTypes.number,
  /**
   * Slide image aspect ratio.
   */
  aspectRatio: PropTypes.number,
  /**
   * Indicates if carousel should start again after last slide.
   */
  cycle: PropTypes.bool,
  /**
   * Make carousel swipeable on mobile.
   */
  swipe: PropTypes.bool,
  /**
   * Make carousel playing immediately after load.
   */
  autoPlay: PropTypes.bool,
  /**
   * Make carousel pausing when hovering a slide.
   */
  pauseOnHover: PropTypes.bool,

  hideControls: PropTypes.bool,
  /**
   * Connects and entitles slide image with caption by adding dynamic aria-labelledby attribute on image.
   */
  ariaLabelledbyPrefix: PropTypes.string,
  /**
   * Add additional components inside a carousel.
   */
  children: PropTypes.oneOfType([childrenPropType, childrenRenderPropType])
};

Carousel.defaultProps = {
  slides: [],
  animationDuration: ANIMATION_DURATION,
  slideDuration: SLIDE_DURATION,
  aspectRatio: ASPECT_RATIO,
  cycle: true,
  swipe: true,
  autoPlay: true,
  pauseOnHover: true
};

export default Carousel;
