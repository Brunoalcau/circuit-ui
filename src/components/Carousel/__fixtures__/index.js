// eslint-disable-next-line import/prefer-default-export
export const SLIDES = [
  {
    image: {
      src: 'https://placedog.net/800',
      alt: 'doggy'
    }
  },
  {
    image: {
      src: 'https://placedog.net/900',
      alt: 'doggy dog'
    }
  },
  {
    image: {
      src: 'https://placedog.net/1000',
      alt: 'woof woof'
    }
  }
];
