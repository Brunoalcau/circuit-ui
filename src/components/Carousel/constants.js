export const ASPECT_RATIO = 1.8;

export const ANIMATION_DURATION = 640;

export const SLIDE_DURATION = 5000;

export const SLIDE_DIRECTIONS = {
  FORWARD: 'forward',
  BACK: 'back'
};
