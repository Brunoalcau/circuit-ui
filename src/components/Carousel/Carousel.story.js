import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';
import { object, number, boolean } from '@storybook/addon-knobs/react';

import { GROUPS } from '../../../.storybook/hierarchySeparators';
import withTests from '../../util/withTests';

import Container from './components/Container';
import Slides from './components/Slides';
import Slide from './components/Slide';
import SlideImage from './components/SlideImage';
import Controls from './components/Controls';
import Buttons from './components/Buttons';
import Status from './components/Status';
import Carousel from './Carousel';
import { ASPECT_RATIO, ANIMATION_DURATION, SLIDE_DURATION } from './constants';

import { SLIDES } from './__fixtures__';

const { ButtonList, NextButton, PrevButton } = Buttons;

// eslint-disable-next-line react/prop-types
const CustomCarousel = ({ slides }) => {
  const total = slides.length;
  const [step, setStep] = useState(0);
  const goBack = () => setStep(step === 0 ? total - 1 : step - 1);
  const goForward = () => setStep(step === total - 1 ? 0 : step + 1);

  return (
    <Container>
      <Slides>
        {slides.map(({ image }, index) => (
          <Slide
            key={index}
            index={index}
            step={step}
            style={{
              opacity: step === index ? 1 : 0,
              transition: 'opacity .3s ease-in'
            }}
          >
            <SlideImage
              src={image.src}
              alt={image.alt}
              aspectRatio={number('Aspect ratio', 2.5)}
            />
          </Slide>
        ))}
      </Slides>
      <Controls>
        <ButtonList>
          <PrevButton onClick={goBack} />
          <Status style={{ marginLeft: 8 }} step={step} total={total} />
          <NextButton onClick={goForward} />
        </ButtonList>
      </Controls>
    </Container>
  );
};

storiesOf(`${GROUPS.COMPONENTS}|Carousel`, module)
  .addDecorator(withTests('Carousel'))
  .add('Carousel: Stateful', () => (
    <div style={{ width: '60vw' }}>
      <Carousel
        slides={object('Slides', SLIDES)}
        slideDuration={number('Slide duration', SLIDE_DURATION)}
        animationDuration={number('Animation duration', ANIMATION_DURATION)}
        aspectRatio={number('Aspect ratio', ASPECT_RATIO)}
        cycle={boolean('Cycle', true)}
        swipe={boolean('Swipe', true)}
        autoPlay={boolean('Auto play', true)}
        pauseOnHover={boolean('Pause on hover', true)}
      />
    </div>
  ))
  .add('Carousel: Composed', () => (
    <div style={{ width: '60vw' }}>
      <CustomCarousel slides={object('Slides', SLIDES)} />
    </div>
  ));
