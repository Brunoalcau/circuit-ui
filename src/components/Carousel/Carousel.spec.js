import React from 'react';

import Carousel from './Carousel';
import { SLIDES } from './__fixtures__';

// enable when migrated to React 16.8
describe.skip('Carousel', () => {
  describe('styles', () => {
    it('should not render without slides data', () => {
      const actual = create(<Carousel />);

      expect(actual).toMatchSnapshot();
    });

    it('should render with default styles', () => {
      const actual = create(<Carousel slides={SLIDES} />);

      expect(actual).toMatchSnapshot();
    });

    it('should render with children as a function', () => {
      const actual = create(
        <Carousel slides={SLIDES}>
          {({ step }) => <h1>Carousel heading for step #{step}</h1>}
        </Carousel>
      );

      expect(actual).toMatchSnapshot();
    });

    it('should render with children as a node', () => {
      const actual = create(
        <Carousel slides={SLIDES}>
          <h1>Carousel heading</h1>
        </Carousel>
      );

      expect(actual).toMatchSnapshot();
    });
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      jest.resetModules();

      const wrapper = renderToHtml(<Carousel slides={SLIDES} />);
      const actual = await axe(wrapper);

      expect(actual).toHaveNoViolations();
    });
  });
});
