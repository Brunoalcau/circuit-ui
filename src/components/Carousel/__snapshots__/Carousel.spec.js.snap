// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel styles should not render without slides data 1`] = `null`;
